packages/internal/identity-resolver/src/identity-resolver.ts

```diff
@@ -50,15 +48,38 @@ export class IdentityResolver {
     return {
       did: document.id,
       pds: new URL(service.serviceEndpoint),
-      handle: extractHandle(document),
+      handle: getValidHandle(document),
     }
   }
 
+  public async getDocument(
+    input: string,
+    options?: ResolveIdentityOptions,
+  ): Promise<DidDocument<AtprotoIdentityDidMethods>> {
+    return isResolvedHandle(input)
+      ? this.getDocumentFromDid(input, options)
+      : this.getDocumentFromHandle(input, options)
+  }
+
   public async getDocumentFromDid(
     did: NonNullable<ResolvedHandle>,
     options?: ResolveDidOptions,
   ): Promise<DidDocument<AtprotoIdentityDidMethods>> {
-    return this.didResolver.resolve(did, options)
+    const document = await this.didResolver.resolve(did, options)
+
+    const handle = getValidHandle(document)
+    if (handle) {
+      options?.signal?.throwIfAborted()
+
+      const did = await this.handleResolver.resolve(handle, options)
+      if (did !== document.id) {
+        throw new TypeError(
```

Comment:
I am not sure `TypeError` is a fit for this case since all the values are used appropriately, it's just the state of the network that doesn't line up.  Not a big deal though.

(I also try to use `TypeError` with care since it is used by libraries distinguishing system errors from application/library/operational errors, as in bounce for example https://github.com/hapijs/bounce/blob/c2ee4d7c71d339e740d89e7f84664a17655ddcb9/lib/index.js#L9-L29)

Reply:
did this to match existing behavior. Adapted to add explicit error classes.